Tools/paletteable sprite editor/js/lib/marvinj-0.9.js

```diff
@@ -0,0 +1,2665 @@
+		var yC=Math.ceil(matrix.length/2);
+		
+		for(var i=0; i<matrix.length; i++){
+			for(var j=0; j<matrix[0].length; j++){
+					
+				if(matrix[i][j] != 0){		
+					nx = x + (j-xC);
+					ny = y + (i-yC);
+					
+					if(nx >= 0 && nx < imageOut.getWidth() && ny >= 0 && ny < imageOut.getHeight()){
+						
+						resultRed	+=	(matrix[i][j]*(imageIn.getIntComponent0(nx, ny)));
+						resultGreen += 	(matrix[i][j]*(imageIn.getIntComponent1(nx, ny)));
+						resultBlue	+=	(matrix[i][j]*(imageIn.getIntComponent2(nx, ny)));
+					}
+					
+					
+				}
+				
+				
+				
+			}
+		}
+		
+		resultRed 	= Math.abs(resultRed);
+		resultGreen = Math.abs(resultGreen);
+		resultBlue = Math.abs(resultBlue);
+		
+		// allow the combination of multiple applications
+		resultRed 	+= imageOut.getIntComponent0(x,y);
+		resultGreen += imageOut.getIntComponent1(x,y);
+		resultBlue 	+= imageOut.getIntComponent2(x,y);
+		
+		resultRed 	= Math.min(resultRed, 255);
+		resultGreen = Math.min(resultGreen, 255);
+		resultBlue 	= Math.min(resultBlue, 255);
+		
+		resultRed 	= Math.max(resultRed, 0);
+		resultGreen = Math.max(resultGreen, 0);
+		resultBlue 	= Math.max(resultBlue, 0);
+		
+		imageOut.setIntColor(x, y, imageIn.getAlphaComponent(x, y), Math.floor(resultRed), Math.floor(resultGreen), Math.floor(resultBlue));
+	};
+
+	function Moravec(){
+		MarvinAbstractImagePlugin.super(this);
+		this.load();
+	}
+
+	Moravec.prototype.load = function(){
+		this.setAttribute("matrixSize", 3);
+		this.setAttribute("threshold", 0);
+	}
+
+	Moravec.prototype.process = function
+	(
+		imageIn, 
+		imageOut,
+		attrOut,
+		mask, 
+		previewMode
+	)
+	{
+		var matrixSize = this.getAttribute("matrixSize");
+		var threshold = this.getAttribute("threshold");
+		
+		var tempImage = new MarvinImage(imageIn.getWidth(), imageIn.getHeight());
+		Marvin.grayScale(imageIn, tempImage);
+		
+		var cornernessMap = MarvinJSUtils.createMatrix2D(tempImage.getWidth(), tempImage.getHeight(), 0);
+		var cornernessMapOut = MarvinJSUtils.createMatrix2D(tempImage.getWidth(), tempImage.getHeight(), 0);
+		
+		for(var y=0; y<tempImage.getHeight(); y++){
+			for(var x=0; x<tempImage.getWidth(); x++){
+				cornernessMap[x][y] = this.c(x,y,matrixSize,tempImage);
+				
+				if(cornernessMap[x][y] < threshold){
+					cornernessMap[x][y] = 0;
+				}
+			}
+		}
+		
+		for(var x=0; x<cornernessMap.length; x++){
+			for(var y=0; y<cornernessMap[x].length; y++){
+				cornernessMapOut[x][y] = this.nonmax(x,y,matrixSize,cornernessMap);
+				
+				if(cornernessMapOut[x][y] > 0){
+					cornernessMapOut[x][y] = 1;
+				}
+			}
+		}
+		
+		if(attrOut != null){
+			attrOut.set("cornernessMap", cornernessMapOut);
+		}
+	}
+	
+	Moravec.prototype.nonmax = function(x, y, matrixSize, matrix){
+		var s = Math.floor(matrixSize/2);
+		if(x-(s+1) >= 0 && x+(s+1) < matrix.length && y-(s+1) >= 0 && y+(s+1) < matrix[0].length){
+			for(var i=-s; i<=s; i++){
+				for(var j=-s; j<=s; j++){
+					if(i != 0 || j != 0){
```

Comment:
Codacy found an issue: [Use ===/!== to compare with true/false or Numbers](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6395467)